src/ray/rpc/client_call.h

```diff
@@ -67,140 +153,228 @@ class ClientCallImpl : public ClientCall {
-  grpc::CompletionQueue cq_;
+  void SetReplyReaderTag(ClientCallTag *reader_tag) { reader_tag_ = reader_tag; }
 
-  /// Polling thread to check the completion queue.
-  std::thread polling_thread_;
+  ClientCallTag *GetReplyReaderTag() override { return reader_tag_; }
+
+ private:
+  /// The reply message.
+  Reply reply_;
+
+  /// The callback function to handle the reply.
+  ClientCallback<Reply> callback_;
+
+  /// Async writer and reader.
+  std::unique_ptr<grpc::ClientAsyncReaderWriter<Request, Reply>> client_stream_;
+
+  ClientCallTag *tag_;
+  /// Reading and writing in client stream call are asynchronous, we should use different
+  /// type of tags to distinguish the call tag got from completion queue in polling
+  /// thread. We should put the reader tag into the completion queue when try to read a
+  /// reply asynchronously in the stream.
+  ClientCallTag *reader_tag_;
+
+  /// Max size of writing buffer, we cann't write next request unless the buffer is not
+  /// full or the completion queue is ready to write.
+  int max_buffer_size_;
+
+  // Indicates whether it's ready to write next request into stream. It's available only
+  // if we receive the writer tag from the completion queue.
+  bool ready_to_write_;
+
+  // Mutex to protect the shared variables in stream call.
+  // Variables `buffer_` and `ready_to_write_` are used in separate threads.
+  std::mutex stream_call_mutex_;
+
+  // Buffer for sending requests to the server. We need write the request into the queue
+  // when the server is not ready to write.
+  std::queue<std::unique_ptr<Request>> buffer_;
+
+  // Reading and writing are diabled unless the connection between client and server has
+  // setup successfully. Use `promise` and `future` to make the `Connect` function a
+  // synchronous call.
+  std::promise<void> connect_promise_;
+
+  bool is_running_;
+
+  friend class ClientCallManager;
+};
+
+/// This class wraps a `ClientCall`, and is used as the `tag` of gRPC's `CompletionQueue`.
+///
+/// The lifecycle of a `ClientCallTag` is as follows.
+///
+/// For a default asynchronous grpc call. When a client submits a new gRPC request, a new
+/// `ClientCallTag` object will be created by `ClientCallMangager::CreateCall`. Then the
+/// object will be used as the tag of `CompletionQueue`.
+///
+/// For a stream grpc call, we need to used two tags to distinguish the type of tags got
+/// from the completion queue. When received a reply reader tag we should handle the reply
+/// from the server. For a default tag, we should keep trace of the state of the client
+/// stream client to connect, send requests and finish the stream.
+class ClientCallTag {
+ public:
+  enum class TagType {
+    /// A tag for common uses, such as connecting, sending requests and sending writes
+    /// done.
+    DEFAULT,
+    /// Only used to read reply from the server. Each client stream call should contains
+    /// one
+    /// reply reader tag to handle reply.
+    REPLY_READER,
+  };
```

Comment:
I'm feeling we don't need the `TagType`, as we already have type in the `ClientCall`.

Reply:
We need tag to distinguish whether the tag from completion queue is for reading or writing